Quantize conv2d transpose

kumasento opened the pull request:
This PR creates a new quantized kernel called `QuantizedDeconv2D`, which performs a deconvolution computation (also known as `conv2d_transpose` in TensorFlow) in a quantized mode.

The current progress of this new kernel includes:

1. A reference implementation of deconvolution
2. Two unit tests with small dimensions and different strides
3. Only `VALID` padding is allowed

Further guidance on this work is welcomed!

kumasento committed:
check whether the key exists before using it

kumasento committed:
resolved conflicts in tensorflow quantize_graph

kumasento committed:
temporary fix for cuDNN building problems

kumasento committed:
passed quantized_deconv_ops_test with very simple version

kumasento committed:
Merge remote-tracking branch 'upstream/master'

kumasento committed:
Merge branch 'master' into quantize_conv2d_transpose

kumasento committed:
updated stride > 1 unit test

kumasento committed:
removed verbose output

kumasento committed:
Merge remote-tracking branch 'upstream/master'

kumasento committed:
Merge branch 'master' into quantize_conv2d_transpose

tensorflow-jenkins commented:
Can one of the admins verify this patch?

kumasento committed:
removed local directory links

kumasento committed:
updated quantize_graph tool

drpngx commented:
@andrewharp what do you think?

petewarden commented on tensorflow/core/kernels/quantized_deconv_ops.cc:
```diff
@@ -0,0 +1,315 @@
+/* Copyright 2015 The TensorFlow Authors. All Rights Reserved.
```
Update to 2017

kumasento replied:
Done

petewarden commented on tensorflow/core/kernels/quantized_deconv_ops.cc:
```diff
@@ -0,0 +1,315 @@
+      // TODO: Remove these checkers when they are ready
+
+      // Stride can only be 1 at the moment
+      // OP_REQUIRES(context, strides_[1] == 1,
```
Can you clarify whether this is still true? If not, remove the commented-out code, otherwise make it active.

petewarden commented on tensorflow/core/ops/nn_ops.cc:
```diff
@@ -2412,6 +2412,51 @@ max_output: The float value that the highest quantized output value represents.
+  .Attr("strides: list(int)")
+  .Attr(GetPaddingAttrString()) // add padding string
+  .SetShapeFn([] (InferenceContext *c) {
+    // TODO: Don't understand this shape handle funtion for now, will check later
```
Please copy the code from the Conv2DBackpropInput shape function here:
https://github.com/kumasento/tensorflow/blob/3d50e2ef0f823b3c2e7dc5070734fe418d4ce32b/tensorflow/core/ops/nn_ops.cc#L566

kumasento replied:
Done

petewarden commented on tensorflow/stream_executor/cuda/cuda_dnn.cc:
```diff
@@ -40,7 +40,7 @@ limitations under the License.
 #include "tensorflow/stream_executor/stream.h"
 #include "tensorflow/stream_executor/stream_executor_pimpl.h"
 // clang-format off
-#include "cuda/include/cudnn.h"
```
Is this change relevant to the PR?

kumasento replied:
Done

petewarden commented on tensorflow/tools/graph_transforms/quantize_nodes.cc:
```diff
@@ -724,6 +732,7 @@ Status QuantizeNodes(const GraphDef& input_graph_def,
         }
         // This isn't a float op, so don't quantize it.
         if (!are_all_float) {
+          LOG(INFO) << SummarizeNodeDef(float_node);
```
Is this change needed?

kumasento replied:
Done

petewarden commented on tensorflow/tools/graph_transforms/quantize_nodes.cc:
```diff
@@ -84,6 +84,13 @@ const std::vector<QuantizedOpInfo>& GetQuantizedOpList() {
+       {{"Tinput", DT_QUINT8}, {"Tfilter", DT_QUINT8}, {"out_type", DT_QINT32}},
+       DT_QUINT8,
+       DT_QINT32,
+       {},
```
Can you also add a simple test to https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/graph_transforms/quantize_nodes_test.cc#L315

petewarden reviewed (commented):
Thanks for putting this together, a very nice contribution! I've added some comments.

kumasento committed:
Merge remote-tracking branch 'upstream/master' into quantize_conv2d_transpose

kumasento committed:
changed copyright from 2015 to 2017

kumasento committed:
copied set shape function from Conv2DBackpropInput

kumasento committed:
removed redundant log

kumasento commented:
@petewarden Hi Pete, thank you so much for your comments and review! I have fixed several of them and will come back later for the rest.

rmlarsen commented:
@petewarden could you take another look, please?

drpngx commented:
@petewarden any cycles to look at this? Feel free to assign someone else.

sb2nov commented:
@petewarden ping

petewarden commented on tensorflow/core/ops/nn_ops.cc:
```diff
@@ -2489,6 +2489,40 @@ max_output: The float value that the highest quantized output value represents.
+  .Attr("Tinput: quantizedtype")
+  .Attr("Tfilter: quantizedtype")
+  .Attr("out_type: quantizedtype = DT_QINT32") // default quantized type
+  // .Attr("Tinput: numbertype")
```
Remove this commented-out code.

petewarden commented on tensorflow/core/ops/nn_ops.cc:
```diff
@@ -2489,6 +2489,40 @@ max_output: The float value that the highest quantized output value represents.
+  .Attr("strides: list(int)")
+  .Attr(GetPaddingAttrString()) // add padding string
+  .SetShapeFn([] (InferenceContext *c) {
+    // TF_RETURN_IF_ERROR(shape_inference::Conv2DShape(c));
```
Remove this commented-out code.

petewarden commented on tensorflow/stream_executor/cuda/cuda_dnn.cc:
```diff
@@ -40,7 +40,7 @@ limitations under the License.
 #include "tensorflow/stream_executor/stream_executor_pimpl.h"
 // clang-format off
-#include "cuda/include/cudnn.h"
+#include "cuda/cudnn.h"
```
Can we revert the changes to this file entirely, since we're not expecting any alterations?

petewarden reviewed (changes requested):
Sorry for the very slow response on this. I've added some minor comments, but overall looks good, thanks.

kumasento committed:
Merge remote-tracking branch 'upstream/master'

kumasento committed:
Merge branch 'master' into quantize_conv2d_transpose

kumasento committed:
Improved numerical stability in quantized deconv

This commit revised the original quantized conv2d transpose
implementation with similar approaches in quantized conv2d.
A bug related to output data initialization is also fixed.
Several other issues mentioned in PR#9924 are also resolved.

kumasento committed:
Renamed to QuantizedConv2DTranspose

QuantizedConv2DBackpropInput is a bad name, because Conv2DBackpropInput
has a different semantic meaning comparing with Conv2DTranspose, which
is actually what I intend to implement.

kumasento commented:
@petewarden Hi Pete, sorry for my late reply. Thank you for your kind comments. I resolved most of your issues mentioned previously.

There is only one thing left: you mentioned that I should implement a test in tensorflow/tools/graph_transforms/quantize_nodes.cc
I tried, but the problem is: there is no corresponding C++ implementation of `Conv2DTranspose`,
and what we have is `Conv2DBackpropInput`. So I think maybe it is a better idea to establish a clear mapping between `QuantizedConv2DTranspose` and `Conv2DBackpropInput` before writing the test. 

However, before I start to work on it, could you please help me with the following question:
Should I build `QuantizedConv2DTranspose` on a new `QuantizedConv2DBackpropInput` op and map that to `Conv2DBackpropInput`, or:
directly map `QuantizedConv2DTranspose` to `Conv2DBackpropInput`?

Thanks again for all your help.

petewarden commented:
Thanks for your flexibility on this!

> directly map QuantizedConv2DTranspose to Conv2DBackpropInput?

I think this approach is best.

kumasento committed:
Add quantize Conv2DBackpropInput test

kumasento committed:
Fixed a bug of quantize mapping

Previously the Conv2DBackpropInput cannot be successfully mapped to
QuantizeConv2DTranpose, now it can, with a slight revision in
QuantizedOpInfo.
Also, I changed the tensor mapping QuantizedConv2DTranpose to fit the
order in Conv2DBackpropInput.

kumasento commented:
@petewarden Hi, I just added a simple test regarding quantizing `Conv2DBackpropInput` to `QuantizedConv2DTranspose`. 
There is also a slight change in the quantize graph generation part, which can allow you convert a node to a quantized one with a custom node name, not just Quantized*.

Please let me know if you have further suggestions. Thanks!

kumasento committed:
Created a stronger test in graph transform

The previous test cannot cover many cases. This revised test is better.

martinwicke commented:
@petewarden can you take another look?

drpngx commented:
@petewarden any luck with this?

tensorflowbutler commented:
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

tensorflowbutler commented:
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

tensorflowbutler commented:
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

tensorflowbutler commented:
Nagging Assignee @petewarden: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

tensorflowbutler commented:
Nagging Assignee @petewarden: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

tensorflowbutler commented:
Nagging Assignee @petewarden: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

xiaocai00 commented:
any luck with this one?

tensorflowbutler commented:
Nagging Assignee @petewarden: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

tensorflowbutler commented:
Nagging Assignee @petewarden: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

tensorflowbutler commented:
Nagging Assignee @petewarden: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

tensorflowbutler commented:
Nagging Assignee @petewarden: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

tensorflowbutler commented:
Nagging Assignee @petewarden: It has been 60 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

tensorflowbutler commented:
Nagging Assignee @petewarden: It has been 75 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

tensorflowbutler commented:
Nagging Assignee @petewarden: It has been 90 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

tensorflowbutler commented:
Nagging Assignee @petewarden: It has been 105 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

petewarden commented:
Sorry for the long delay on this PR. We're moving away from this approach for supporting quantization inside TensorFlow (all quantized ops now live in TF Lite), so unfortunately we'll need to close this request.
